File: docs/caching/hetero_micro_bench.lua
#!/usr/bin/env lua
-- Micro-benchmark: heterogeneous dispatch (Lua version)
-- Compares dispatch caching across Lisp/Scheme implementations

local CACHE_SIZE = 8

local DispatchCache = {}
DispatchCache.__index = DispatchCache

function DispatchCache:new()
    local self = setmetatable({}, DispatchCache)
    self.entries = {}
    self.next_idx = 0
    self.hits = 0
    self.misses = 0
    return self
end

function DispatchCache:lookup(key)
    for i = 1, CACHE_SIZE do
        local entry = self.entries[i]
        if entry and entry[1] == key then
            return entry[2]
        end
    end
    return nil
end

function DispatchCache:insert(key, fn)
    self.entries[self.next_idx + 1] = {key, fn}
    self.next_idx = (self.next_idx + 1) % CACHE_SIZE
end

local function clause_number(x)
    return {"number", x}
end

local function clause_string(x)
    return {"string", #x}
end

local function clause_table(x)
    return {"table", #x}
end

local function clause_boolean(x)
    return {"boolean", x}
end

local function clause_other(x)
    return {"other", x}
end

local hetero_cache = DispatchCache:new()

local function get_type_key(x)
    local t = type(x)
    if t == "number" then
        return "number"
    elseif t == "string" then
        return "string"
    elseif t == "table" then
        return "table"
    elseif t == "boolean" then
        return "boolean"
    else
        return "other"
    end
end

local function dispatch_hetero_uncached(x)
    local t = type(x)
    if t == "number" then
        return {"number", x}
    elseif t == "string" then
        return {"string", #x}
    elseif t == "table" then
        return {"table", #x}
    elseif t == "boolean" then
        return {"boolean", x}
    else
        return {"other", x}
    end
end

local function dispatch_hetero_cached(x)
    local key = get_type_key(x)
    local hit = hetero_cache:lookup(key)

    if hit then
        hetero_cache.hits = hetero_cache.hits + 1
        return hit(x)
    end

    hetero_cache.misses = hetero_cache.misses + 1
    local t = type(x)
    if t == "number" then
        hetero_cache:insert(key, clause_number)
        return clause_number(x)
    elseif t == "string" then
        hetero_cache:insert(key, clause_string)
        return clause_string(x)
    elseif t == "table" then
        hetero_cache:insert(key, clause_table)
        return clause_table(x)
    elseif t == "boolean" then
        hetero_cache:insert(key, clause_boolean)
        return clause_boolean(x)
    else
        hetero_cache:insert(key, clause_other)
        return clause_other(x)
    end
end

-- Test data: 200,000 calls with 5-type repeating cycle
local test_data = {}
local cycle = {1, "test string", {1, 2, 3, 4, 5}, true, "symbol"}
for i = 1, 200000 do
    table.insert(test_data, cycle[((i - 1) % 5) + 1])
end

local function benchmark_uncached(iterations)
    local times = {}
    for run = 1, iterations do
        local result = 0
        local start_time = os.clock()
        for _, item in ipairs(test_data) do
            local r = dispatch_hetero_uncached(item)
            if r then result = result + 1 end
        end
        local elapsed = (os.clock() - start_time) * 1000.0
        print(string.format("  Run %d: %.1f seconds", run, elapsed / 1000.0))
        table.insert(times, result)
    end
    return times
end

local function benchmark_cached(iterations)
    local times = {}
    for run = 1, iterations do
        hetero_cache.hits = 0
        hetero_cache.misses = 0
        local result = 0
        local start_time = os.clock()
        for _, item in ipairs(test_data) do
            local r = dispatch_hetero_cached(item)
            if r then result = result + 1 end
        end
        local elapsed = (os.clock() - start_time) * 1000.0
        print(string.format("  Run %d: %.1f seconds", run, elapsed / 1000.0))
        table.insert(times, result)
    end
    return times
end

local function run_all_benchmarks()
    print("\n================================")
    print("Lua Heterogeneous Dispatch Caching Micro-Benchmark")
    print("================================")
    print(string.format("Implementation: Lua %s", _VERSION))
    print("Test data: 200,000 calls over repeating 5-type cycle")
    print("  Type cycle: number -> string -> table -> boolean -> other\n")

    print("Warming up interpreter (10,000 calls)...")
    for i = 1, 10000 do
        dispatch_hetero_uncached(test_data[(i % 200000) + 1])
    end
    for i = 1, 10000 do
        hetero_cache.hits = 0
        hetero_cache.misses = 0
        dispatch_hetero_cached(test_data[(i % 200000) + 1])
    end
    print("Warmup complete.\n")

    print("=== Uncached Dispatch (3 iterations) ===")
    benchmark_uncached(3)

    print("\n=== Cached Dispatch (3 iterations) ===")
    benchmark_cached(3)

    print("\nCached Dispatch Stats:")
    local total = hetero_cache.hits + hetero_cache.misses
    print(string.format("  Cache hits: %d", hetero_cache.hits))
    print(string.format("  Cache misses: %d", hetero_cache.misses))
    if total > 0 then
        local hit_rate = 100.0 * hetero_cache.hits / total
        print(string.format("  Hit rate: %.4f%%", hit_rate))
    end

    print("\n================================")
    print("Benchmark Complete")
    print("================================")
end

run_all_benchmarks()

File: docs/caching/hetero_micro_bench.js
#!/usr/bin/env node
// Micro-benchmark: heterogeneous dispatch (JavaScript version)
// Compares dispatch caching across implementations

const CACHE_SIZE = 8;

class DispatchCache {
  constructor() {
    this.entries = new Array(CACHE_SIZE);
    this.nextIdx = 0;
    this.hits = 0;
    this.misses = 0;
  }

  lookup(key) {
    for (let i = 0; i < CACHE_SIZE; i++) {
      const entry = this.entries[i];
      if (entry && entry[0] === key) {
        return entry[1];
      }
    }
    return null;
  }

  insert(key, fn) {
    this.entries[this.nextIdx] = [key, fn];
    this.nextIdx = (this.nextIdx + 1) % CACHE_SIZE;
  }
}

function clauseNumber(x) {
  return ["number", x];
}

function clauseString(x) {
  return ["string", x.length];
}

function clauseObject(x) {
  return ["object", Object.keys(x).length];
}

function clauseBoolean(x) {
  return ["boolean", x];
}

function clauseOther(x) {
  return ["other", x];
}

const heteroCache = new DispatchCache();

function getTypeKey(x) {
  if (typeof x === "number") {
    return "number";
  } else if (typeof x === "string") {
    return "string";
  } else if (typeof x === "object" && x !== null) {
    return "object";
  } else if (typeof x === "boolean") {
    return "boolean";
  } else {
    return "other";
  }
}

function dispatchHeteroUncached(x) {
  if (typeof x === "number") {
    return ["number", x];
  } else if (typeof x === "string") {
    return ["string", x.length];
  } else if (typeof x === "object" && x !== null) {
    return ["object", Object.keys(x).length];
  } else if (typeof x === "boolean") {
    return ["boolean", x];
  } else {
    return ["other", x];
  }
}

function dispatchHeteroCached(x) {
  const key = getTypeKey(x);
  const hit = heteroCache.lookup(key);

  if (hit) {
    heteroCache.hits++;
    return hit(x);
  }

  heteroCache.misses++;
  const t = typeof x;
  if (t === "number") {
    heteroCache.insert(key, clauseNumber);
    return clauseNumber(x);
  } else if (t === "string") {
    heteroCache.insert(key, clauseString);
    return clauseString(x);
  } else if (t === "object" && x !== null) {
    heteroCache.insert(key, clauseObject);
    return clauseObject(x);
  } else if (t === "boolean") {
    heteroCache.insert(key, clauseBoolean);
    return clauseBoolean(x);
  } else {
    heteroCache.insert(key, clauseOther);
    return clauseOther(x);
  }
}

// Test data: 200,000 calls with 5-type repeating cycle (repeated 10x to increase workload)
const testData = [];
const cycle = [1, "test string", { a: 1, b: 2, c: 3, d: 4, e: 5 }, true, "symbol"];
for (let repeat = 0; repeat < 10; repeat++) {
  for (let i = 0; i < 200000; i++) {
    testData.push(cycle[i % 5]);
  }
}

function benchmarkUncached(iterations) {
  const times = [];
  for (let run = 1; run <= iterations; run++) {
    const startTime = performance.now();
    for (let i = 0; i < testData.length; i++) {
      const r = dispatchHeteroUncached(testData[i]);
      if (r) {
        globalResult = (globalResult + 1) % 1000000; // Force side effect
      }
    }
    const elapsed = performance.now() - startTime;
    console.log(`  Run ${run}: ${(elapsed / 1000).toFixed(1)} seconds`);
    times.push(elapsed);
  }
  return times;
}

function benchmarkCached(iterations) {
  const times = [];
  for (let run = 1; run <= iterations; run++) {
    heteroCache.hits = 0;
    heteroCache.misses = 0;
    const startTime = performance.now();
    for (let i = 0; i < testData.length; i++) {
      const r = dispatchHeteroCached(testData[i]);
      if (r) {
        globalResult = (globalResult + 1) % 1000000; // Force side effect
      }
    }
    const elapsed = performance.now() - startTime;
    console.log(`  Run ${run}: ${(elapsed / 1000).toFixed(1)} seconds`);
    times.push(elapsed);
  }
  return times;
}

let globalResult = 0;

function runAllBenchmarks() {
  console.log("\n================================");
  console.log("JavaScript Heterogeneous Dispatch Caching Micro-Benchmark");
  console.log("================================");
  console.log(`Implementation: Node.js ${process.version}`);
  console.log("Test data: 2,000,000 calls over repeating 5-type cycle (10× iterations)");
  console.log("  Type cycle: number -> string -> object -> boolean -> other\n");

  console.log("Warming up interpreter (10,000 calls)...");
  for (let i = 0; i < 10000; i++) {
    dispatchHeteroUncached(testData[i % 200000]);
  }
  for (let i = 0; i < 10000; i++) {
    heteroCache.hits = 0;
    heteroCache.misses = 0;
    dispatchHeteroCached(testData[i % 200000]);
  }
  console.log("Warmup complete.\n");

  console.log("=== Uncached Dispatch (10 iterations) ===");
  benchmarkUncached(10);

  console.log("\n=== Cached Dispatch (10 iterations) ===");
  benchmarkCached(10);

  console.log("\nCached Dispatch Stats:");
  const total = heteroCache.hits + heteroCache.misses;
  console.log(`  Cache hits: ${heteroCache.hits}`);
  console.log(`  Cache misses: ${heteroCache.misses}`);
  if (total > 0) {
    const hitRate = (100.0 * heteroCache.hits) / total;
    console.log(`  Hit rate: ${hitRate.toFixed(4)}%`);
  }

  console.log(`\nInternal state (preventing JIT optimization): ${globalResult}`);
  console.log("\n================================");
  console.log("Benchmark Complete");
  console.log("================================");
}

runAllBenchmarks();
